### PR TITLE
Google Ngram based features

### PR DESCRIPTION
**Pull request by sbhaktha**
@mhrmm : Here it is. Although only using unigrams (Nodes dataset) of the Google Ngram data, utility class has been coded to be general (if we want to use bigram/trigram based features in the future).

For any given token, the word associated with the token is looked up in the Google Ngram frequency map, filtered to just the ones matching the token's POS tag. Then a normalized frequency distribution of the various dependency labels for that word and that POS tag is used to generate the features.

### CODE REVIEW

**Comment by sbhaktha** [comment]
Parse scores:

Baseline (withou Google Ngram):

```
[info] UAS: 3450 / 4209 = 81.97%
[info] UAS (no punc): 3033 / 3678 = 82.46%
[info] LAS: 3211 / 4209 = 76.29%
[info] LAS (no punc): 2800 / 3678 = 76.13%
[info] Path Accuracy (labeled, full): 2630 / 4209 = 62.49%
[info] Path Accuracy (unlabeled, full): 2928 / 4209 = 69.57%
[info] Path Accuracy (labeled, ignorePunc): 2241 / 3678 = 60.93%
[info] Path Accuracy (unlabeled, ignorePunc): 2531 / 3678 = 68.81%
[info] Cpos Tagging: 3919 / 4209 = 93.11%
[info] Parsed 401 sentences in 3.7 seconds, an average of 109.1 sentences per second.
[info] Checking oracle accuracy on test set question-bank/original/stan3-4-ch-dependencies/qbank.test.400.conllx.
```

With Google Ngram features:

```
[info] UAS: 3475 / 4209 = 82.56%
[info] UAS (no punc): 3058 / 3678 = 83.14%
[info] LAS: 3259 / 4209 = 77.43%
[info] LAS (no punc): 2846 / 3678 = 77.38%
[info] Path Accuracy (labeled, full): 2710 / 4209 = 64.39%
[info] Path Accuracy (unlabeled, full): 2959 / 4209 = 70.30%
[info] Path Accuracy (labeled, ignorePunc): 2317 / 3678 = 63.00%
[info] Path Accuracy (unlabeled, ignorePunc): 2562 / 3678 = 69.66%
[info] Cpos Tagging: 3916 / 4209 = 93.04%
[info] Parsed 401 sentences in 4.1 seconds, an average of 98.9 sentences per second.
[info] Checking oracle accuracy on test set question-bank/original/stan3-4-ch-dependencies/qbank.test.400.conllx.
```

Shows a small improvement, though not that stable if you run it 3 times and take average. :)

**Comment by sbhaktha** [comment]
@mhrmm Added POS tag based features as well. Scores are frustratingly in the same range for all combinations of features I tried.

Verbnet + Google PosTag + Google DepLabel

```
[info] UAS: 8820 / 10427 = 84.59%
[info] UAS (no punc): 7757 / 9130 = 84.96%
[info] LAS: 7527 / 10427 = 72.19%
[info] LAS (no punc): 7353 / 9130 = 80.54%
[info] Path Accuracy (labeled, full): 6229 / 10427 = 59.74%
[info] Path Accuracy (unlabeled, full): 7593 / 10427 = 72.82%
[info] Path Accuracy (labeled, ignorePunc): 6128 / 9130 = 67.12%
[info] Path Accuracy (unlabeled, ignorePunc): 6601 / 9130 = 72.30%
[info] Cpos Tagging: 9807 / 10427 = 94.05%
[info] Parsed 1001 sentences in 15.6 seconds, an average of 64.3 sentences per second.
```


Verbnet + Google PosTag

```
[info] UAS: 8792 / 10427 = 84.32%
[info] UAS (no punc): 7748 / 9130 = 84.86%
[info] LAS: 7479 / 10427 = 71.73%
[info] LAS (no punc): 7325 / 9130 = 80.23%
[info] Path Accuracy (labeled, full): 6285 / 10427 = 60.28%
[info] Path Accuracy (unlabeled, full): 7651 / 10427 = 73.38%
[info] Path Accuracy (labeled, ignorePunc): 6181 / 9130 = 67.70%
[info] Path Accuracy (unlabeled, ignorePunc): 6656 / 9130 = 72.90%
[info] Cpos Tagging: 9755 / 10427 = 93.56%
[info] Parsed 1001 sentences in 13.5 seconds, an average of 74.2 sentences per second.
```


Verbnet + Google DepLabel

```
[info] UAS: 8871 / 10427 = 85.08%
[info] UAS (no punc): 7807 / 9130 = 85.51%
[info] LAS: 7545 / 10427 = 72.36%
[info] LAS (no punc): 7368 / 9130 = 80.70%
[info] Path Accuracy (labeled, full): 6254 / 10427 = 59.98%
[info] Path Accuracy (unlabeled, full): 7659 / 10427 = 73.45%
[info] Path Accuracy (labeled, ignorePunc): 6136 / 9130 = 67.21%
[info] Path Accuracy (unlabeled, ignorePunc): 6657 / 9130 = 72.91%
[info] Cpos Tagging: 9815 / 10427 = 94.13%
[info] Parsed 1001 sentences in 14.1 seconds, an average of 70.8 sentences per second.
```


No Verbnet, Google PosTag + Google DepLabel

```
[info] UAS: 8878 / 10427 = 85.14%
[info] UAS (no punc): 7800 / 9130 = 85.43%
[info] LAS: 7593 / 10427 = 72.82%
[info] LAS (no punc): 7387 / 9130 = 80.91%
[info] Path Accuracy (labeled, full): 6273 / 10427 = 60.16%
[info] Path Accuracy (unlabeled, full): 7635 / 10427 = 73.22%
[info] Path Accuracy (labeled, ignorePunc): 6138 / 9130 = 67.23%
[info] Path Accuracy (unlabeled, ignorePunc): 6620 / 9130 = 72.51%
[info] Cpos Tagging: 9818 / 10427 = 94.16%
[info] Parsed 1001 sentences in 16.2 seconds, an average of 61.7 sentences per second.
```


No Verbnet, Google PosTag

```
[info] UAS: 8771 / 10427 = 84.12%
[info] UAS (no punc): 7731 / 9130 = 84.68%
[info] LAS: 7479 / 10427 = 71.73%
[info] LAS (no punc): 7325 / 9130 = 80.23%
[info] Path Accuracy (labeled, full): 6152 / 10427 = 59.00%
[info] Path Accuracy (unlabeled, full): 7536 / 10427 = 72.27%
[info] Path Accuracy (labeled, ignorePunc): 6051 / 9130 = 66.28%
[info] Path Accuracy (unlabeled, ignorePunc): 6548 / 9130 = 71.72%
[info] Cpos Tagging: 9783 / 10427 = 93.82%
[info] Parsed 1001 sentences in 13.3 seconds, an average of 75.5 sentences per second.
```

No Verbnet, Google DepLabel

```
[info] UAS: 8822 / 10427 = 84.61%
[info] UAS (no punc): 7739 / 9130 = 84.76%
[info] LAS: 7535 / 10427 = 72.26%
[info] LAS (no punc): 7336 / 9130 = 80.35%
[info] Path Accuracy (labeled, full): 6247 / 10427 = 59.91%
[info] Path Accuracy (unlabeled, full): 7567 / 10427 = 72.57%
[info] Path Accuracy (labeled, ignorePunc): 6122 / 9130 = 67.05%
[info] Path Accuracy (unlabeled, ignorePunc): 6553 / 9130 = 71.77%
[info] Cpos Tagging: 9793 / 10427 = 93.92%
[info] Parsed 1001 sentences in 14.1 seconds, an average of 71.2 sentences per second.
```

No Verbnet, No Google

```
[info] UAS: 8744 / 10427 = 83.86%
[info] UAS (no punc): 7716 / 9130 = 84.51%
[info] LAS: 7385 / 10427 = 70.83%
[info] LAS (no punc): 7247 / 9130 = 79.38%
[info] Path Accuracy (labeled, full): 6232 / 10427 = 59.77%
[info] Path Accuracy (unlabeled, full): 7683 / 10427 = 73.68%
[info] Path Accuracy (labeled, ignorePunc): 6145 / 9130 = 67.31%
[info] Path Accuracy (unlabeled, ignorePunc): 6704 / 9130 = 73.43%
[info] Cpos Tagging: 9743 / 10427 = 93.44%
[info] Parsed 1001 sentences in 9.7 seconds, an average of 102.9 sentences per second.
```

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala [hunk 1]
```diff
@@ -161,9 +164,106 @@ case class VerbnetTagger(verbnet: Verbnet) extends SentenceTransform {
...
+          }
+        }).groupBy(_._1).mapValues(_.flatMap(v => v._2))
+        untagged.updateProperties(frequencyFeatureMap)
+    }))
+  }
+}
+
+/** Frequency Distribution of POS tags for words based on Google Ngram's Nodes (unigrams).
+  */
+case class GoogleUnigramPostagTagger(googleNgram: GoogleNGram) extends SentenceTransform {
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: There's a lot of code duplication between this and the previous class. Is there a way to refactor?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala [hunk 2]
```diff
@@ -161,9 +164,106 @@ case class VerbnetTagger(verbnet: Verbnet) extends SentenceTransform {
...
+          tok.word.name, googleNgram.ngramMap, googleNgram.frequencyCutoff
+        )
+        // Create feature for each dependency label based on the normalized frequency
+        // bucket it lies in.
+        val frequencyFeatureMap: Map[Symbol, Set[Symbol]] = (for {
+          postag <- postagFreqMap.keySet
+        } yield {
+          val normalizedFrequency = postagFreqMap(postag)
+          val symbolSetWithCurrentPostag = Set(Symbol(postag))
+          if (normalizedFrequency <= 0.0009) {
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: Probably you don't want this many buckets. It likely just confuses the learner to have to treat 50-60% completely separately from 60-70%. I would suggest a bucketing more like the following: 1-5% 5-20% 20-50% 50-95% 95+%.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+import scala.io._
+
+/** A class that parses Google N-Gram data
+  * (http://commondatastorage.googleapis.com/books/syntactic-ngrams/index.html) to provide
+  * information about a requested n-gram.
+  * Takes the datastore location details for a data directory and parses each file, expected
+  * to be in the following format
+  * (from https://docs.google.com/document/d/14PWeoTkrnKk9H8_7CfVbdvuoFZ7jYivNTkBX2Hj7qLw/edit) -
+  *   head_word<TAB>syntactic-ngram<TAB>total_count<TAB>counts_by_year
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala]: You'll need to add a `format: OFF` to this to preserve the indentations.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  *   “word/pos-tag/dep-label/head-index”.
+  * The word field can contain any non-whitespace character.
+  * The other fields can contain any non-whitespace character except for ‘/’.
+  *   pos-tag is a Penn-Treebank part-of-speech tag.
+  *   dep-label is a stanford-basic-dependencies label.
+  *   head-index is an integer, pointing to the head of the current token.
+  *   “1” refers to the first token in the list, 2 the second,
+  *      and 0 indicates that the head is the root of the fragment.
+  */
+case class GoogleNGram(
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala]: Rename to DatastoreGoogleNgram?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+/** Utility case classes to represent information associated with an ngram in the Google Ngram
+  * corpus.
+  */
+case class SyntacticInfo(word: String, posTag: String, depLabel: String, headIndex: Int)
+case class NgramInfo(syntacticNgram: Seq[SyntacticInfo], frequency: Long)
+
+/** Companion object. Contains methods to parse a Google Ngram corpus. This is not specific to
+  * the type of corpus, i.e. whether unigram, bigram, etc.
+  */
+object GoogleNGram {
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala]: But keep this object name as "GoogleNGram" (just move the serializer out into a DatastoreGoogleNGram companion object).

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+  /** Looks up specified ngramMap for the given word and returns a map of the frequency for each
+    * POS tag for the given word, normalized over the total frequency for all possible POS tags.
+    * @param word the word to look up
+    * @param ngramMap the table mapping a word to the sequence of NgramInfos, as obtained from the
+    * GoogleNGram class object
+    * @param frequencyCutoff the frequency cutoff that was used to construct the map. This is used
+    * here to shift the scale of the frequencies to start from the cutoff point instead of 1.
+    */
+  def getPosTagNormalizedDistribution(
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala]: Also quite a bit of code duplication between this and the previous function.

**Comment by mhrmm** [comment]
First pass done. A couple comments.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala (hunk 1 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: Yeah. I thought we might experiment with different buckets for the two and just left them there. I'll change it.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala (hunk 2 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: I initially tried with 25%, 50%, 75% and 100%. I'll put them back in and retry.

Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTransform.scala]: Sorry I read your suggested bucketing. Putting that in and trying.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala (hunk 6 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala]: Actually they are not the same- they pass different first augments to `getTokenUnigramInfo`. The later 'for` comprehensions are also doing different things.

**Comment by sbhaktha** [comment]
Addressed comments. Added config parameter to choose feature(s) to use from Google Ngram: depLabel/posTag.

**Comment by mhrmm** [comment]
LGTM

**Comment by mhrmm** [comment]
BTW, did you upload the filtered version of the google unigrams to the datastore?

**Comment by sbhaktha** [comment]
Yeah, to use it, you need the following in the `featuretaggers.config` file:

```
googleUnigram {
  group: "org.allenai.nlp.resources"
  name: "googleNgramsNodes-20130501-freq1000Filtered"
  version: 1
  features: [ "depLabel", "posTag" ]
}
```

They are all in a single ~1.3 gig file. I tested with it and seems to work fine.

**Comment by sbhaktha** [comment]
@mhrmm Changed default to filtered dataset. Build passed. But Shippable took atrociously long-- close to an hour to finish running all the tests because of the long d/l time from datastore. The file's about 1.3 gigs, not sure why it takes this long. I can do some things to cut down the file size even more. Look for another commit from me!

**Comment by sbhaktha** [comment]
@mhrmm : Done! Just changed the file in the datastore. It's down to 5.6 megs. Build time has come down from an hour to 16 mins.

**Pull request merged**